diff --git a/progs/ghosts.v b/progs/ghosts.v
index 601ae95e..1b2d2731 100644
--- a/progs/ghosts.v
+++ b/progs/ghosts.v
@@ -1100,7 +1100,7 @@ Lemma all_disjoint_rev : forall l, all_disjoint l <-> all_disjoint (rev l).
 Proof.
   split; [apply all_disjoint_rev1|].
   intros ?%all_disjoint_rev1.
-  rewrite rev_involutive in H0; auto.
+  rewrite rev_involutive in *; auto.
 Qed.
 
 Lemma  maps_add_rev : forall l, all_compatible l -> maps_add (rev l) = maps_add l.
