diff --git a/Remakefile.in b/Remakefile.in
index 85ef334..0237d89 100644
--- a/Remakefile.in
+++ b/Remakefile.in
@@ -144,7 +144,7 @@ public: deps.png deps.map html/index.html
 	cp -r html deps.png public/
 
 install:
-	dir="$DESTDIR@COQUSERCONTRIB@/Interval"
+	dir="${DESTDIR}@COQUSERCONTRIB@/Interval"
 	mkdir -p $dir
 	for d in Eval Float Integral Interval Missing Poly Real; do mkdir -p $dir/$d; done
 	for f in $(FILES); do cp src/$f.vo $dir/$f.vo; done
