From bc53f843f90c0d78d2121c0a529cfb3fae39bfa0 Mon Sep 17 00:00:00 2001
From: Michael Soegtrop <msoegtrop@apple.com>
Date: Tue, 19 Oct 2021 11:04:56 +0200
Subject: [PATCH] Allow Coq 8.14.0

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index c8fe3898..666df46a 100644
--- a/Makefile
+++ b/Makefile
@@ -21,7 +21,7 @@ COQLIB=$(shell $(COQC) -where | tr -d '\r' | tr '\\' '/')
 
 # Check Coq version
 
-COQVERSION= 8.12.0 or-else 8.12.1 or-else 8.12.2 or-else 8.13+beta1 or-else 8.13.0 or-else 8.13.1 or-else 8.13.2
+COQVERSION= 8.12.0 or-else 8.12.1 or-else 8.12.2 or-else 8.13+beta1 or-else 8.13.0 or-else 8.13.1 or-else 8.13.2 or-else 8.14.0
 
 COQV=$(shell $(COQC) -v)
 ifneq ($(IGNORECOQVERSION),true)
-- 
2.30.1 (Apple Git-130)

