From 5facdb7fd256a5b3efe8c805f2666d613f302bd3 Mon Sep 17 00:00:00 2001
From: Michael Soegtrop <7895506+MSoegtropIMC@users.noreply.github.com>
Date: Sat, 4 Jul 2020 18:42:50 +0200
Subject: [PATCH] Windows build: use architecture dependent version of windres

---
 Makefile.ide | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/Makefile.ide b/Makefile.ide
index 640ee7b188..789acee5ec 100644
--- a/Makefile.ide
+++ b/Makefile.ide
@@ -296,10 +296,12 @@ $(COQIDEAPP):$(COQIDEAPP)/Contents/Resources
 # CoqIde for Windows special targets
 ###########################################################################
 
+# This is either x86_64-w64-mingw32 or i686-w64-mingw32
+TARGET_ARCH=$(shell $(CC) -dumpmachine)
+
 %.o: %.rc
 	$(SHOW)'WINDRES    $<'
-	$(HIDE)i686-w64-mingw32-windres -i $< -o $@
-
+	$(HIDE)$(TARGET_ARCH)-windres -i $< -o $@
 
 # For emacs:
 # Local Variables:
-- 
2.21.0

