From b2f8e4d0fdea4611b56f8e2c2feade3e8fb632b7 Mon Sep 17 00:00:00 2001
From: Michael Soegtrop <7895506+MSoegtropIMC@users.noreply.github.com>
Date: Sun, 27 Dec 2020 15:18:54 +0100
Subject: [PATCH] Allow dev version of Menhir

---
 configure | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/configure b/configure
index 7cbb9d7d..4c494379 100755
--- a/configure
+++ b/configure
@@ -556,7 +556,7 @@ fi
 
 MENHIR_REQUIRED=20190626
 echo "Testing Menhir... " | tr -d '\n'
-menhir_ver=`menhir --version 2>/dev/null | sed -n -e 's/^.*version \([0-9]*\).*$/\1/p'`
+menhir_ver=$(menhir --version 2>/dev/null | sed -n -E -e 's/^.*version (unreleased|[0-9]*).*$/\1/p')
 case "$menhir_ver" in
   20[0-9][0-9][0-9][0-9][0-9][0-9])
       if test "$menhir_ver" -ge $MENHIR_REQUIRED; then
@@ -576,6 +576,18 @@ case "$menhir_ver" in
           echo "Error: CompCert requires a version greater or equal to $MENHIR_REQUIRED."
           missingtools=true
       fi;;
+  unreleased)
+      echo "version $menhir_ver -- acceptable!"
+      menhir_dir=$(ocamlfind query menhirLib 2>/dev/null) || \
+      menhir_dir=$(menhir --suggest-menhirLib) || \
+      menhir_dir=""
+      menhir_dir=$(echo "$menhir_dir" | tr -d '\r' | tr '\\' '/')
+      if test ! -d "$menhir_dir"; then
+          echo "Error: cannot determine the location of the Menhir API library."
+          echo "This can be due to an incorrect Menhir package."
+          echo "Consider using the OPAM package for Menhir."
+          missingtools=true
+      fi;;
   *)
         echo "NOT FOUND"
         echo "Error: make sure Menhir version $MENHIR_REQUIRED or later is installed."
-- 
2.29.2

