From 3893f766d4d09f713b4bbdf74dd5e70f46dc7ca5 Mon Sep 17 00:00:00 2001
From: Jim Fehrle <jim.fehrle@gmail.com>
Date: Fri, 21 May 2021 14:02:10 -0700
Subject: [PATCH] Overlay for Coq PR 14354

---
 src/coq_elpi_utils.ml | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/coq_elpi_utils.ml b/src/coq_elpi_utils.ml
index 359ae29..3f802bd 100644
--- a/src/coq_elpi_utils.ml
+++ b/src/coq_elpi_utils.ml
@@ -6,7 +6,7 @@ module API = Elpi.API
 
 let of_coq_loc l = {
   API.Ast.Loc.source_name =
-    (match l.Loc.fname with Loc.InFile x -> x | Loc.ToplevelInput -> "(stdin)");
+    (match l.Loc.fname with Loc.InFile {file} -> file | Loc.ToplevelInput -> "(stdin)");
   source_start = l.Loc.bp;
   source_stop = l.Loc.ep;
   line = l.Loc.line_nb;
@@ -18,7 +18,7 @@ let to_coq_loc {
   line_starts_at = line_starts_at;
   source_start = source_start;
   source_stop = source_stop;
-} = Loc.create (Loc.InFile source_name) line line_starts_at source_start source_stop
+} = Loc.create (Loc.InFile {dirpath=None; file=source_name}) line line_starts_at source_start source_stop
 
 let err ?loc msg =
   let loc = Option.map to_coq_loc loc in
-- 
2.30.1 (Apple Git-130)

